substrate/frame/balances/src/lib.rs

```diff
@@ -768,12 +769,19 @@ pub mod pallet {
 						"account with a non-zero reserve balance has no provider refs, account_id: '{:?}'.",
 						who
 					);
-					a.free = a.free.max(Self::ed());
+					a.free = a.free.max(ed);
 					system::Pallet::<T>::inc_providers(who);
 				}
 				let _ = system::Pallet::<T>::inc_consumers_without_limit(who).defensive();
 			}
-			// Should never fail - we're only setting a bit.
+
+			// Ensure the account holds at least ED of free balance.
+			let free = a.free;
+			if free < ed && a.reserved >= ed - free {
+				a.free = ed;
+				a.reserved -= ed - free;
```

Comment:
We cannot just change the reserve of an account, reserved balances are GUARANTEED to be slashable.  
To me it looks like this should reap accounts that are below ED. Contract accounts would then need to be air-dropped... which is a pain.  
Alternative is to just "give" them `ed` for free. But Total issuance needs to be updated AFAIK.

Reply:
I don't get why not to fix the `did_provide` logic (like https://github.com/paritytech/polkadot-sdk/pull/2700/commits/2f847113a22bd08bca53179fdbe296afc7041fb0) rather than faffing about changing balances which comes with a whole new set of considerations and risks?

Reply:
This is not correct the invariant on the new logic is that `account.free >= Self::ed()`
This previous fix was breaking some other tests. I can lookup which one later on to let you know.

Reply:
I guess what we could try is to lower the provider_count by one knowing that it will be bumped by 1 later on, not sure if that's a safe approach or not